### PR TITLE
Dev console and UI for listing running dev services

### PR DESCRIPTION
**Pull request by ozangunalp**
- Refactoring for the dev services to produce build items even though a running dev service already exists. 
- Dev console commands In dev mode: [c] for getting the status of running containers, [g] for following container logs on the console
- Dev UI for listing services and injected configs.

Screenshot 2022-01-20 at 08 57 19

Screenshot 2022-01-20 at 08 58 03

### CODE REVIEW

**Comment by gsmet** [comment]
Nice work!

I haven't looked at the code but here are a couple of comments on the UI:
- can you make them sorted alphabetically? (the order in the screenshot looks a bit random)
- any chance you could add a thing with the full command to connect to the service when it makes sense? Typically the `psql` command line to connect to the database.

**Comment by Ladicek** [comment]
This seems like a decent start of a solution for #22943.

**Comment by ozangunalp** [comment]
@Ladicek 👍 I didn't see #22943 before starting to work on this, but it's the same idea.
@gsmet Sorting definitely makes sense! 

I'll look to open a shell on the container from dev console.
I didn't want to add any runtime dependencies (ie. to docker-client), that's why dev UI is somewhat limited. 
My first idea was to have a log and shell console on the dev UI.

And I'll try to find a place to add some tests...

**Comment by gsmet** [comment]
@ozangunalp I was not talking about opening a shell, just providing the command line you can use to connect from your host when it makes sense. So it's just about formatting a string with the proper info.

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/DevServicesConfigResultBuildItem.java
Comment (mkouba) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/DevServicesConfigResultBuildItem.java:14]: Pls add `since()` and `forRemoval()` if needed.

File: extensions/datasource/deployment-spi/src/main/java/io/quarkus/datasource/deployment/spi/DevServicesDatasourceResultBuildItem.java
Comment (mkouba) [inline extensions/datasource/deployment-spi/src/main/java/io/quarkus/datasource/deployment/spi/DevServicesDatasourceResultBuildItem.java:10]: Is it ok to remove this class completely? I mean it's public and part of the `spi` package so I'd consider it to be part of the public API...

**Review by mkouba — COMMENTED** [review]
I've added some minor comments. The result (Dev UI) looks ok but I can't speak about the code and the logic behind. We'll need someone who worked on the dev services... 

CC @stuartwdouglas @gsmet @geoand @Sanne

File: extensions/datasource/deployment-spi/src/main/java/io/quarkus/datasource/deployment/spi/DevServicesDatasourceResultBuildItem.java
Comment (ozangunalp) [reply extensions/datasource/deployment-spi/src/main/java/io/quarkus/datasource/deployment/spi/DevServicesDatasourceResultBuildItem.java:10]: You are right on that. There was no usage on quarkus repo and external extensions but I found at least 2 usages on github.

Comment (geoand) [reply extensions/datasource/deployment-spi/src/main/java/io/quarkus/datasource/deployment/spi/DevServicesDatasourceResultBuildItem.java:10]: As @mkouba says, we cant remove build items that are part of the SPI

File: extensions/mongodb-client/runtime/pom.xml
Comment (geoand) [inline extensions/mongodb-client/runtime/pom.xml:34]: Just to be totally clear, this does not result in the inclusion of any additional dependencies in prod mode, right?

**Review by geoand — COMMENTED** [review]
This is very neat work! 

I added a couple comments

File: extensions/mongodb-client/runtime/pom.xml
Comment (ozangunalp) [reply extensions/mongodb-client/runtime/pom.xml:34]: Yes, the runtime module (`quarkus-devservices`) only depends on `quarkus-arc`.

Comment (geoand) [reply extensions/mongodb-client/runtime/pom.xml:34]: Great, thanks for clarifying!

**Review by stuartwdouglas — COMMENTED** [review]
LGTM, although that failures looks kinda related (possibly an existing memory leak that is exposed by the new test, unless this PR actually adds a new memory leak)

File: extensions/devservices/deployment/src/main/java/io/quarkus/devservices/deployment/DevServicesProcessor.java
Comment (stuartwdouglas) [inline extensions/devservices/deployment/src/main/java/io/quarkus/devservices/deployment/DevServicesProcessor.java:135]: If you flip this if condition (i.e. check containerIds.isEmpty() first) it will fix the memory leak. It is also a better design, as we should not be initializing the docker client if the project does not actually have any docker based dev services.

This is an existing leak that only affects QuarkusUnitTest (so is not something that really affects end users). Basically the testcontainers docker client is never closed, and it hangs onto the class loader forever. This change means that every test in the devmode module now opens a docker client, and this is never cleaned up so it runs out of memory.

We should look at some way of fixing this (possibly making testcontainers parent first), but it is an existing issue so does not have to happen with this PR (if you added a dep that used dev services to this module it would fail the same way).

**Review by stuartwdouglas — DISMISSED** [review]

File: extensions/devservices/deployment/src/main/java/io/quarkus/devservices/deployment/DevServicesProcessor.java
Comment (ozangunalp) [reply extensions/devservices/deployment/src/main/java/io/quarkus/devservices/deployment/DevServicesProcessor.java:135]: Thanks for the explanation @stuartwdouglas. This totally makes sense now.

File: extensions/devservices/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (gsmet) [inline extensions/devservices/runtime/src/main/resources/META-INF/quarkus-extension.yaml:9]: I'm not sure to understand why we need a runtime extension here?

Do we have anything runtime for the other Dev UI things?

**Review by gsmet — DISMISSED** [review]
Added a small comment and a question.

File: extensions/devservices/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (ozangunalp) [reply extensions/devservices/runtime/src/main/resources/META-INF/quarkus-extension.yaml:9]: Dev UI section is easier to add this way. Otherwise, I'd have to add it to vertx-http module

File: extensions/apicurio-registry-avro/runtime/pom.xml
Comment (ozangunalp) [reply extensions/apicurio-registry-avro/runtime/pom.xml:57]: 👍 I'll check the others

File: extensions/devservices/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (gsmet) [reply extensions/devservices/runtime/src/main/resources/META-INF/quarkus-extension.yaml:9]: @stuartwdouglas do we really want some Dev UI related components to leak into the runtime? From my point of view, it's really something we should avoid if we can but given you designed this thing, I'll let the decision to you.

Comment (ozangunalp) [reply extensions/devservices/runtime/src/main/resources/META-INF/quarkus-extension.yaml:9]: As I wrote to @geoand s earlier comment, only dependency devservices runtime module has is quarkus-arc.

Comment (stuartwdouglas) [reply extensions/devservices/runtime/src/main/resources/META-INF/quarkus-extension.yaml]: I have pushed a commit that removes the runtime dep, and moves the components into various other deployment modules. The 'Dev Services' link now lives on the first 'Configuration' card.

**Comment by ozangunalp** [comment]
@stuartwdouglas thanks for this! I started doing exactly this to not add a new module but I couldn't render things inside dev UI without a runtime class.

**Comment by stuartwdouglas** [comment]
@gsmet I think your concerns have been addressed

**Comment by ozangunalp** [comment]
@gsmet @stuartwdouglas Do you have anything else in mind before we merge this?

**Comment by gsmet** [comment]
I dismissed my review as my concerns have been addressed. @phillip-kruger maybe you could have a look at this one from the UI side? Either now or after merging?

**Comment by phillip-kruger** [comment]
This is very nice ! I can have a look at the UI, but maybe after merge...

**Comment by gsmet** [comment]
Yes, probably better after merge. I'll let @stuartwdouglas approve and merge this one.

**Review by stuartwdouglas — APPROVED** [review]

**Comment by phillip-kruger** [comment]
see https://github.com/quarkusio/quarkus/pull/23636 for a potential screen update.

**Pull request merged**
